### ISSUE TITLE
Provide picture tag name with configurable srcset media and sizes attributes

### ISSUE THREAD

**Issue by gforcada** [plone/plone.namedfile#111]
The browser views to generate image scales here in `plone.namedfile` can generate `<img>` tags with a `srcset` attribute for 2x and 3x versions of the very same image.

i.e.

```html
<img 
    src="http://localhost:8080/Plone/bla/@@images/c062d71e-0b44-4592-a577-2641d4253fb5.jpeg"
    alt="Description"
    title="more description"
    height="800"
    width="1200"
    srcset="
        http://localhost:8080/Plone/bla/@@images/HASH-FOR-2x.jpeg 2x,
        http://localhost:8080/Plone/bla/@@images/HASH-FOR-3x.jpeg 3x" />
```

On the current project I'm working on we need much more advanced requirements, something like:

```html
<picture>
  <source
      media="(min-width: 1024px)"
      sizes="600px"
      srcset="https://dummyimage.com/360x3:4 360w,
              https://dummyimage.com/560x3:4 560w,
              https://dummyimage.com/768x3:4 768w,
              https://dummyimage.com/960x3:4 960w,
              https://dummyimage.com/1240x3:4 1240w,
              https://dummyimage.com/1440x3:4 1440w,
              https://dummyimage.com/1680x3:4 1680w,
              https://dummyimage.com/1920x3:4 1920w">
  <source
      media="(min-width: 520px)"
      sizes="100vw"
      srcset="https://dummyimage.com/360x3:2 360w,
              https://dummyimage.com/560x3:2 560w,
              https://dummyimage.com/768x3:2 768w,
              https://dummyimage.com/960x3:2 960w,
              https://dummyimage.com/1240x3:2 1240w,
              https://dummyimage.com/1440x3:2 1440w,
              https://dummyimage.com/1680x3:2 1680w,
              https://dummyimage.com/1920x3:2 1920w">
  <source
      media="(min-width: 0)"
      sizes="100vw"
      srcset="https://dummyimage.com/360x3:4 360w,
              https://dummyimage.com/560x3:4 560w,
              https://dummyimage.com/768x3:4 768w,
              https://dummyimage.com/960x3:4 960w,
              https://dummyimage.com/1240x3:4 1240w,
              https://dummyimage.com/1440x3:4 1440w,
              https://dummyimage.com/1680x3:4 1680w,
              https://dummyimage.com/1920x3:4 1920w">
  <img
      class="CUSTOM-IMG-CSS-CLASS"
      sizes="(min-width: 1024px) 600px, 100vw"
      srcset="https://dummyimage.com/360x3:2 360w,
              https://dummyimage.com/560x3:2 560w,
              https://dummyimage.com/768x3:2 768w,
              https://dummyimage.com/960x3:2 960w,
              https://dummyimage.com/1240x3:2 1240w,
              https://dummyimage.com/1440x3:2 1440w,
              https://dummyimage.com/1680x3:2 1680w,
              https://dummyimage.com/1920x3:2 1920w"
      src="https://dummyimage.com/768x3:2"
      alt=""
      decoding="async"
      loading="lazy">
</picture>
```

i.e. given a few media queries (the `media` attribute on the `source` tags) that go hand in hand (I'm not totally sure about that though) with the `sizes` attribute (again on the `source` tag), generate a few scales (that might be a list of the usual scales names that we have already in Plone) and set for each of them a specific *condition descriptor* (as [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset) calls them).

Put a fallback (the `img` tag on the example above) for ensuring that something gets displayed on that [~5% that does not support srcset](https://caniuse.com/?search=srcset).

Has anyone tried to do something this elaborated so far by (over)using the current `plone.namedfile` facilities? 🤔

**Comment by mauritsvanrees** [comment]
Nice to see that more people are thinking about this! I have given thought to it this week as well. Let me paste some notes.

I was looking at which code is involved in giving an img tag in a rich text field its final content. Mostly this is **`plone.outputfilters`**.

You have the `html_to_plone_outputfilters_html` transform which is registered in `portal_transforms`. It has a short [`convert` method](https://github.com/plone/plone.outputfilters/blob/4.1.0/plone/outputfilters/transforms/html_to_plone_outputfilters_html.py#L28-L35). It does `getAdapters((context, request), IFilter)` and runs each filter over the original html. This is `plone.outputfilters.interfaces.IFilter`.

The only `IFilter` in core Plone is the `resolve_uid_and_caption` filter from `plone.outputfilters` itself, [registered here](https://github.com/plone/plone.outputfilters/blob/4.1.0/plone/outputfilters/filters/configure.zcml#L6-L11). Its most [important lines are these](https://github.com/plone/plone.outputfilters/blob/4.1.0/plone/outputfilters/filters/resolveuid_and_caption.py#L183-L187). Basically: find all `img` tags in the html, for each image resolve the image to an actual persistent/stable/uuid-like scale, and set the new src attribute.

Sample original img tag, as literally found in the raw value of a textfield:

```
<img alt="" class="image-richtext image-inline" data-linktype="image" data-scale="preview"
 data-val="061605ac7eb64b6aa9259106ff4ba7c0" height=""
 src="../resolveuid/061605ac7eb64b6aa9259106ff4ba7c0/@@images/image/preview"
 width=""/>
```

After the src has been replaced:

```
<img alt="" class="image-richtext image-inline" data-linktype="image" data-scale="preview"
 data-val="061605ac7eb64b6aa9259106ff4ba7c0" height=""
 src="http://localhost:8080/Plone/dscn0368.jpg/@@images/01e8df62-286a-4551-a5ec-074320fd4c49.jpeg"
 width=""/>
```

If we want a `picture` tag instead, what we could do, is create another `IFilter`. It looks for `img` tags. For each img tag it looks up the full image by UID from `data-val`. And it may take the scale name from `data-scale` into consideration. Then it wraps the img tag in a picture tag and adds source tags. Maybe the img tag could be changed to contain as src the original full image, if we think this is better than the current tag. For the source tags: perhaps add a source for each scale that is smaller than the requested one (preview here).

This should be doable in a small add'on in current Plone 5.2 or 6.0 without changes. Or as new feature in `plone.outputfilters`. To make it nicer, we could add a registry setting with the responsive image scale breakpoints, and use those when generating the source tags. If we improve the image control in TinyMCE, then instead of getting `data-scale="preview"`, we might get something like `data-size="small/medium/large"` or `data-percentage="50"` and `data-align="left/right/center/filled"`, similar to Volto, and base the generated sources on that. Or see [@MrTango 's idea](https://community.plone.org/t/image-scales-srcset-picture-source-oh-my/14920/4?u=mauritsvanrees).

Note: `plone.outputfilters` says that you can add an `order` attribute on your filter, and the code would call the adapters in the correct order, which could be good for this. I don't see this in the code though: it looks like you would get the adapters in a random order. There is only one so it is okay for now. Should be trivial to add support for this though.

An alternative could be to do this as part of the **transform chain**, so the chain that includes for example the `Diazo/plone.app.theming` transform and the `plone.protect` auto csrf transform. But that sounds trickier, maybe mostly because this would normally use `lxml` instead of `BeautifulSoup`. I would say this transform chain is the best place if you want to do this for *all* img tags, so also the tags outside of the rich text fields. You could still restrict it to looking in the `#content` area if you want.

A second alternative is to do this in **`plone.namedfile`**. Then it would be for *all* images that are called via `@@images`, either in rich text fields or in page templates.

Since we see three different levels where support for this could be implemented, it would be good to have some shared code, perhaps in `plone.scale`. But picking the best place here is tricky.

The question is: **what you put in the sizes?** How do you determine what makes sense? If the img tag is in a template: should be no problem: you know how big the container is where the image is shown, so how large it could grow as maximum.
But if the image is in rich text, then what? I think neither TinyMCE nor any filter/transform knows how wide the text is going to be: when it is the main text of a document, it will be full width, except that you may have left/right portlets on some screen sizes. But the richtext can be in a tile or portlet, which means it could have any width.

Well, I guess that is where small/medium/large comes into play, or even the old scale names. The editor knows he is editing the richtext of a tile, and knows that this tile is for example shown in one of three columns on desktop and in one column on mobile, to put it simply. `size=small` seems best then. And in the main text of a page, the editor would choose `large`, or maybe `medium` if half of the page on desktop is fine.

And then in the code or a control panel we would need a mapping from SML to a list of sizes. See also Maik's code. And then we generate image scales with the widths that best fit this, and add them as srcsets.
Likely, we would generate different widths for small images than for large images. Same/similar number of sources though.

**Comment by MrTango** [comment]
Nice summary, looks like we have a plan :).
I would put the mappings in the image control panel and use it in TinyMCE UI for now. Tiny would still insert a single image with the fallback scale, but we can later do the transform an that and also use the mapping.

**Comment by gforcada** [comment]
Wow, I was mostly "well, the designer gave me this 1000 lines of code for a single image, let's figure out how to do so" and there's so much ongoing work on that, and, indeed, is so much of a complex topic, that I'm so glad that you all are already thinking about it and coming to some solutions already! 🚀 ✨ 🤩 

I would love to be helpful in this effort, my main use case is not the tinymce one, but rather the template one, just so you know 😃

**Comment by flipmcf** [comment]
@mauritsvanrees nailed it above.  

We also use the `resolve_uid_and_caption` filter to change how images are rendered.  
But this will only work with richtext fields.   If you have to do anything outside of rendering a text field, then you have to create a transformer and add it to the transformchain.

SELFLESS PLUG OF MY LAST TALK: https://2021.ploneconf.org/schedule/plone-outputfilters-and-portal-transforms

I really like how outputfilters does it, because you can actually create a browser view and template to handle the image html.  And then use layers for those `special snowflakes` content.

We also went even further with outputfilters and registered different mimetypes and transforms so images can get different html depending on the request.   This is how our site switches between <img> for 'normal' html and <amp-img> for google AMP (special snowflake again)  But again, this is only available at the field level - fields with mimetypes - which is an annoying limitation when you discover how powerful it is.

**Comment by mauritsvanrees** [comment]
> SELFLESS PLUG OF MY LAST TALK: https://2021.ploneconf.org/schedule/plone-outputfilters-and-portal-transforms

Selfless plug of my summary of your talk. :-)
https://maurits.vanrees.org/weblog/archive/2021/10/michael-mcfadden-plone-outputfilters-and-transformchain

**Comment by MrTango** [comment]
Can we talk today about this in the classic-ui channel?
I'm there the whole day sprinting and would like to make some progress here.

**Comment by Rudd-O** [comment]
What @mauritsvanrees is saying is the way to go.

I would be happy to discuss this.  Sorry I missed the message a few hours ago.

**Comment by Rudd-O** [comment]
@flipmcf perhaps the prudent thing to do is to excise the _image attributes_ handling stuff out of `resolveuid_and_caption` (I worked on this code a few weeks ago, it was **horrible** to reason about), and put that in a separate IFilter function, leaving only the resolveuid functionality in the preexisting function.  That way we / @MrTango can deal with the image sizes / srcset stuff separately and cleanly.

**Comment by Rudd-O** [comment]
Here is a meta discussion maybe worth having.

Fundamentally, when I insert a *content* image, I'm not in the slightest interested in pixel size.  What I want is:

* What percentage of the page width or height will the image take?  (In 90% of the cases for my uses it has been something like either 95% height of viewport, or simply as wide as it'll go.  But other uses such as float to the right exist.  I handle these with CSS styles.)
* Will the image appear blurry on mobile devices, or when zoomed in, or across different aspect ratios and screen DPIs?

`srcset` has a great potential to solve the second bullet point for me.

**Comment by MrTango** [comment]
check this out, https://github.com/plone/plone.outputfilters/pull/49
Next step would be to update plone.namedfile to respect the srcset definitions too.
I guess we then can remove some of the existing 2x/3x checkboxes from the control pnael, as you can and should to that inside the sourceset in the media attribute.

**Comment by flipmcf** [comment]
> excise the _image attributes_ handling stuff out of `resolveuid_and_caption`

Wow, actually having a separate `resolveuid` and `caption` ?   I love it.    Any function with an '_and_' in it's name is just advertising the fact it breaks 'single mindedness'.

I did think about this a long time ago, and found that there was a good reason both parts were in the same outputfilter.   I can't remember why.  It's probably an optimization hack.

In principle I agree.  In practice I think it's a bit more complicated.   And I can't remember exactly why I avoided this.  I had a very good reason which I can't remember at all.  But I was younger and naive then, maybe a solution is obvious now?
